File: java/src/main/java/dev/agaber/sports/testing/FakeClock.java
package dev.agaber.sports.testing;

import java.time.Clock;
import java.time.Instant;
import java.time.LocalDate;
import java.time.ZoneId;

public final class FakeClock extends Clock {
  private final Instant now;
  private final ZoneId zoneId;

  public FakeClock(Instant now, ZoneId zoneId) {
    this.now = now;
    this.zoneId = zoneId;
  }

  public FakeClock(Instant now) {
    this(now, ZoneId.systemDefault());
  }

  public FakeClock(LocalDate currentDate) {
    this(currentDate.atStartOfDay(ZoneId.systemDefault()).toInstant());
  }

  @Override
  public ZoneId getZone() {
    return zoneId;
  }

  @Override
  public Clock withZone(ZoneId zoneId) {
    return new FakeClock(now, zoneId);
  }

  @Override
  public Instant instant() {
    return now;
  }
}

File: kotlin/src/main/kotlin/dev/agaber/sports/testing/FakeClock.kt
package dev.agaber.sports.testing

import java.time.Clock
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId

class FakeClock(
  private val now: Instant,
  private val zoneId: ZoneId = ZoneId.systemDefault(),
) : Clock() {
  constructor(now: LocalDate) : this(now.atStartOfDay(ZoneId.systemDefault()).toInstant())

  override fun instant(): Instant = now

  override fun withZone(zone: ZoneId): Clock = FakeClock(now, zone)

  override fun getZone() = zoneId
}
